### PR TITLE
Creates Torch-friendly Event class & adds Stream tracking to autograd

### PR DESCRIPTION
**Pull request by mruberry**
This PR removes the last blocker for #8354 by creating a generic Event class that can be used from Torch code without macros and tracking generic Streams in autograd.

The Event class is designed to dispatch to the appropriate backend. Currently the only backend supported is CUDA, but other devices will likely need a similar interface in the future. It is tested with a new cpp test in cuda_stream_test.

### CODE REVIEW

**Comment by mruberry** [comment]
@zdevito who suggested abstracting Events & Streams for autograd, and @ezyang for HIP issues.

HIP test failures are real and I'll follow-up on them soon.

**Comment by mruberry** [comment]
@zdevito who suggested abstracting Events & Streams for autograd, and @ezyang for HIP issues.

HIP test failures are real and I'll follow-up on them soon.

**Comment by ezyang** [comment]
If you want reviews mark as reviewers ;)

**Comment by ezyang** [comment]
What happened with the `type_and_shape.h` header?

File: aten/src/ATen/cuda/detail/CUDAHooks.h
Comment (ezyang) [inline aten/src/ATen/cuda/detail/CUDAHooks.h]: We may not need to change anything for this PR, but if we do want to add support for other APIs in the future, CUDAHooks is not a great place to put the stubs, as there is no promise of API uniformity for hooks across device backends. You'll want something more like DeviceGuardImpl, but for events and streams.

File: aten/src/ATen/cuda/detail/CUDAHooks.cpp
Comment (ezyang) [inline aten/src/ATen/cuda/detail/CUDAHooks.cpp]: FYI this one is in DeviceGuardImpl already

File: aten/src/ATen/detail/CUDAHooksInterface.h
Comment (ezyang) [inline aten/src/ATen/detail/CUDAHooksInterface.h]: nit: remove me

File: c10/core/Event.h
Comment (ezyang) [inline c10/core/Event.h]: Ick.

File: c10/core/Event.h
Comment (ezyang) [inline c10/core/Event.h]: Don't you need to know what the device of the Event is at construction time? We don't have to actually make this work for multiple backends to start, but we should at least make sure the API is right (and not hard code CUDA).

File: c10/core/Event.h
Comment (ezyang) [inline c10/core/Event.h]: Is it possible to make this a `unique_ptr` with a function pointer deleter instead? It seems marginally more safe.

File: c10/core/Event.h
Comment (ezyang) [inline c10/core/Event.h]: Ugh, layering violation XD (I am not sure if we've merged libraries to the point where this is tolerated behavior. In particular, I suspect that this change here will break mobile in fbcode, because I'm pretty sure you're referring to a struct that is not synced to xplat. This is not a deep problem, because you're never going to use Event on mobile, but it's something that mechanically will need to be fixed.)

**Comment by ezyang** [comment]
I'm generally OK with the overall direction this patch is going, but as the person who wrote many of our device-agnostic wrappers (Stream, DeviceGuard), I have a few quibbles with the details of this patch.

Overall, there is one design principle I'd like to see in the code here: No direct references to CUDA/HIP specific code in the generic class (Stream, DeviceGuard). Instead, you should either (1) not define any methods which require this (Stream, which is just a dumb wrapper around data, and there are other APIs which handle it) or (2) define your own hook and registry for the functionality you need (that's DeviceGuardImpl for DeviceGuard). My preference is for you to extend DeviceGuardImpl with the new hooks you need (making it even less of a "DeviceGuard" impl thing, but we can always rename it later.)

This might help a bit with your HIP woes, so it might be worth frontloading this work.

**Review by ezyang — CHANGES_REQUESTED** [review]
HIP needs to work, and I think if you restructure per my suggestion it will be better.

**Comment by mruberry** [comment]
Thanks for the review, Ed! Let me rework it with the pattern you're recommending and see if I can smooth out the other rough edges, too.

File: c10/core/Event.h
Comment (zdevito) [inline c10/core/Event.h]: This isn't initialized in the move constructor which means that a destructor can see an unitialized flags struct.

File: c10/core/Event.h
Comment (zdevito) [inline c10/core/Event.h]: What are these flags supposed to be? Should we have an enum here of what is valid?

File: c10/core/Event.h
Comment (zdevito) [inline c10/core/Event.h]: `moveHelper` -> `swap`

File: aten/src/ATen/test/cuda_stream_test.cpp
Comment (zdevito) [inline aten/src/ATen/test/cuda_stream_test.cpp]: I suppose our lower level APIs are already this way, but I find `stream.wait(event)` much clearer than `event.block(stream)`. I am not sure what are reasons are for having the `block` but not the the `wait`.

File: c10/core/Event.h
Comment (zdevito) [inline c10/core/Event.h]: To future proof, this should call the no-args constructor.

File: c10/core/Event.h
Comment (mruberry) [reply c10/core/Event.h]: We could allow (or force) the user to specify the backend when the Event is constructed for "type checking," but it's not strictly necessary. Per Zach's point below we could abstract event flags in an enum to make them backend-agnostic and then lazily construct events on the appropriate backend with the appropriate flags. I think generalizing the flags and optionally allowing the backend to be specified will be the best of both worlds. Sound good?

File: c10/core/Event.h
Comment (mruberry) [reply c10/core/Event.h]: Good catch. Yes, we can use a unique_ptr with a custom deleter here.

File: c10/core/Event.h
Comment (mruberry) [reply c10/core/Event.h]: This can be fixed by exposing the void* pointer. That will also get rid of the friend declaration.

File: c10/core/Event.h
Comment (mruberry) [reply c10/core/Event.h]: I'll get rid of this. See comment below (two birds one stone).

File: aten/src/ATen/cuda/detail/CUDAHooks.h
Comment (mruberry) [reply aten/src/ATen/cuda/detail/CUDAHooks.h]: I'll update the pattern to be analogous to DeviceGuard(Impl).

**Comment by mruberry** [comment]
> What happened with the type_and_shape.h header?

It was improperly removed and is now empty. type_and_shape.h had already been replaced by input_metadata.h, since the latter includes type and shape and more. I noticed while updating input_metadata.h so made sure to delete the empty file.

File: c10/core/Event.h
Comment (mruberry) [reply c10/core/Event.h]: See above comment. Offering a menu (enum) of backend agnostic flags that are translated appropriately sounds like a good idea.

File: aten/src/ATen/test/cuda_stream_test.cpp
Comment (mruberry) [reply aten/src/ATen/test/cuda_stream_test.cpp]: I think we can add stream.wait(event) in this PR, too. Pseudocode: 

`Stream::wait(event) = event.block(this)`

Comment (mruberry) [reply aten/src/ATen/test/cuda_stream_test.cpp]: Added this (tested in next test).

File: aten/src/ATen/cuda/detail/CUDAHooks.cpp
Comment (mruberry) [reply aten/src/ATen/cuda/detail/CUDAHooks.cpp]: Updated. See the change in input_metadata.h.

File: c10/core/Event.h
Comment (mruberry) [reply c10/core/Event.h]: Now initializing in move constructor.

File: c10/core/Event.h
Comment (mruberry) [reply c10/core/Event.h]: Added enum of flags.

File: c10/core/Event.h
Comment (mruberry) [reply c10/core/Event.h]: Made it call a constructor using properties of other.

**Comment by mruberry** [comment]
@zdevito you now have stream.wait(event)
@ezyang I altered the design to build on top of DeviceGuardImplInterface

I think the remaining build failures are unrelated. They appear to be happening in other PRs submitted at the time the last commit to this PR was, and the causes of failure (like a poor SSL connection or DLLs being improperly registered on Windows) don't seem related to this PR.

To recap, this PR creates a generic Event class on top of the backends registered by DeviceGuardImplInterface. That interface was extended and the CUDA and HIP backends updated with event functionality. Other backends received stubs - @ezyang, would you prefer they throw?

Per review, flags were refactored into their own enum and stream.wait(event) was added to the Stream API. Unfortunately, changing the void* event_ pointer to a unique_ptr seems like more of a pain than its worth. Each design either added conceptual or actual overhead.

Two new tests are added for the Event class and stream.wait() functionality, one using Event and one using an InlineEvent directly. No tests currently verify the streams recorded in input_metadata are working properly, but this functionality will be used by the streaming backwards PR.

**Comment by ezyang** [comment]
Pro-tip: don't forget to rerequest review. I get PR review requests notifications more quickly than general CC notifications

File: c10/core/impl/DeviceGuardImplInterface.h
Comment (ezyang) [inline c10/core/impl/DeviceGuardImplInterface.h]: Can this briefly explain why this is a `void**` rather than just a `void*`?

File: c10/core/impl/InlineEvent.h
Comment (ezyang) [inline c10/core/impl/InlineEvent.h]: So, when do we get `struct CUDAEvent : InlineEvent<CUDAGuardImpl>` ;)

File: c10/core/impl/InlineEvent.h
Comment (ezyang) [inline c10/core/impl/InlineEvent.h]: I guess, the reason making this a `unique_ptr` sucked was because you don't statically know the type of the `cudaEvent_t` you're pointing to.

Ignoring the question of unique_ptr versus raw pointer, it might be good to pass it in as a template argument in any case. One mistake I made with InlineDeviceGuard was that it doesn't accept CUDAStream as a template argument; which is why I had to copy paste code in `aten/src/ATen/hip/impl/HIPGuardImplMasqueradingAsCUDA.h` instead of just saying `using OptionalHIPGuardMasqueradingAsCUDA = c10::impl::InlineOptionalDeviceGuard<HIPGuardImplMasqueradingAsCUDA>`. The extra types help you give refined signatures in inline guard which give you the option of not making a wrapper struct. (I still like writing the wrapper struct boilerplate explicitly because it helps people understand what API they can actually use.) But we don't have to handle that in this PR.

Comment (ezyang) [reply c10/core/impl/InlineEvent.h]: And it doesn't really matter here because you haven't added a CUDAEvent class in this PR.

File: c10/core/Stream.h
Comment (ezyang) [inline c10/core/Stream.h]: I'm not convinced `EventInterface` is necessary here. Can't you just make this templated, so that it works on `Event` or `InlineEvent`?

**Comment by ezyang** [comment]
This looks good. The one thing I am curious about is if we actually need `EventInterface`. The rest LGTM (but do read my comments).

File: c10/core/impl/DeviceGuardImplInterface.h
Comment (mruberry) [reply c10/core/impl/DeviceGuardImplInterface.h]: record() may modify the void* event_ pointer since it might allocated a cuda or hip event. So calling it requires &event_ which is of type void**.

File: c10/core/Event.h
Comment (mruberry) [reply c10/core/Event.h]: Sounds good.

File: c10/core/impl/InlineEvent.h
Comment (mruberry) [reply c10/core/impl/InlineEvent.h]: I didn't want to collide with ATen's CUDAEvent, which contains additional backend-specific functionality.

File: c10/core/impl/InlineEvent.h
Comment (mruberry) [reply c10/core/impl/InlineEvent.h]: That's right, we don't know the type of the event. It could be a cudaEvent_t, a hipEvent_t, or some as yet unknown future event type.

Let me think about your idea of using templating to improve on a void*.

File: c10/core/Stream.h
Comment (mruberry) [reply c10/core/Stream.h]: Yes we can. I'll find a different home for the enum.

**Comment by mruberry** [comment]
> This looks good. The one thing I am curious about is if we actually need `EventInterface`. The rest LGTM (but do read my comments).

Thanks Ed! Good suggestions. I'll get them fixed up.

File: c10/core/impl/InlineEvent.h
Comment (mruberry) [reply c10/core/impl/InlineEvent.h]: Unfortunately, I don't think there's much we can do to remove the void* here. The DeviceGuardImplInterface would need to be templatized, too, or the event functions would need to be removed from that interface. VirtualGuardImpl relies on it, however, and the entire point of VirtualGuardImpl is that it can select a backed at runtime.

**Review by ezyang — APPROVED** [review]
Good to go when ci passes

**Comment by mruberry** [comment]
The build failure with the XLA branch is real and needs a small fix. After that we should be good.

**Comment by mruberry** [comment]
Failure of caffe2-py2-cuda9.0-cudnn7-windows-build is occurring on other PRs, too, and seems unrelated. See https://github.com/pytorch/pytorch/pull/24847, for example.

**Comment by mruberry** [comment]
caffe2-py2-cuda9.0-cudnn7-windows-build now passes, although pr/pytorch-win-ws2016-cuda9-cudnn7-py3 is mysteriously failing. I think this errors are unrelated to this PR, however. Landing now.

**Comment by mruberry** [comment]
Blocked on internal test failures. I'll wrap the HIP event destructor in a try-catch block to resolve them.

**Comment by ezyang** [comment]
@mruberry FYI if you're going to be bouncing between fbcode and OSS in the future, using ghstack/ghimport will make this workflow a lot more pleasant.

**Comment by mruberry** [comment]
@ezyang Thanks for the tip, Ed! I'll check it out. (Still some lint issues to resolve.)

**Pull request closed without merging**
